modules/unsupported/elasticsearch/src/main/java/org/geotools/data/elasticsearch/FilterToElastic.java

```diff
@@ -1217,11 +1220,26 @@ private void visitLiteralGeometry(Literal expression) throws IOException {
             final CoordinateSequence coordinates = linearRing.getCoordinateSequence();
             currentGeometry = factory.createLineString(coordinates);
         }
-
-        final String geoJson = GeoJSONWriter.toGeoJSON(currentGeometry);
+        int maxDecimals = getMaxDecimalsForGeometry(currentGeometry);
+        final String geoJson = GeoJSONWriter.toGeoJSON(currentGeometry, maxDecimals);
         currentShapeBuilder = mapReader.readValue(geoJson);
     }
 
+    private int getMaxDecimalsForGeometry(Geometry currentGeometry) {
+        Envelope envelope = currentGeometry.getEnvelopeInternal();
+        double min = Math.min(Math.abs(envelope.getWidth()), Math.abs(envelope.getHeight()));
+        if (min == 0) {
+            LOGGER.log(
+                    Level.WARNING,
+                    "BBox Geometry has no width or height, it is either a point or a line.");
+            return 0;
+        } else if (min >= 0.000001) {
+            return JtsModule.DEFAULT_MAX_DECIMALS;
+        } else {
+            return Math.abs((int) Math.floor(Math.log10(min)) + 1);
+        }
```

Comment:
When one zooms out a lot the numbers start to get silly in the other direction, e.g., longitudes at "10000", for which this algorithm would state something like... 4 decimals? If the log is negative, use it as decimals, but if it's positive, I'd say settle on zero decimals?

This method could be taking an Envelope, be static, and be tested against various situations in JUnit, making sure it works well in a variety of situations, since it's the core of the fix.

Reply:
fixed